### PR TITLE
Introduce @SubclassOptInRequired to codebase and stabilize part of SerialDescriptor's API

### PR DESCRIPTION
**Pull request by sandwwraith**

### CODE REVIEW

**Comment by sandwwraith** [comment]
@qwwdfsad, although I'm not sure if we should introduce it now since https://youtrack.jetbrains.com/issue/KTIJ-30831 may lead to a poor UX

File: core/commonMain/src/kotlinx/serialization/descriptors/SerialDescriptor.kt [hunk 1]
```diff
@@ -275,19 +292,17 @@ public interface SerialDescriptor {
...
      */
-    @ExperimentalSerializationApi
     public fun isElementOptional(index: Int): Boolean
 }
 
 /**
  * Returns an iterable of all descriptor [elements][SerialDescriptor.getElementDescriptor].
  */
-@ExperimentalSerializationApi
 public val SerialDescriptor.elementDescriptors: Iterable<SerialDescriptor>
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/serialization/descriptors/SerialDescriptor.kt:306]: While we are here, let's add a short sample of code?

Comment (qwwdfsad) [reply core/commonMain/src/kotlinx/serialization/descriptors/SerialDescriptor.kt:306]: The same goes for all the other "stable" parts

File: core/commonMain/src/kotlinx/serialization/descriptors/SerialKinds.kt [hunk 2]
```diff
@@ -239,7 +234,7 @@ public sealed class StructureKind : SerialKind() {
  * bounded and sealed polymorphism common property: not knowing the actual type statically and requiring
  * formats to additionally encode it.
  */
-@ExperimentalSerializationApi
+@ExperimentalSerializationApi // Intentionally left experimental to sort out things with buildSerialDescriptor(PolymorphicKind.SEALED)
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/serialization/descriptors/SerialKinds.kt:237]: Thanks for the comment!

File: core/commonMain/src/kotlinx/serialization/descriptors/SerialDescriptor.kt (hunk 1 above)
Comment (sandwwraith) [reply core/commonMain/src/kotlinx/serialization/descriptors/SerialDescriptor.kt:409]: I've added examples for `SerialDescriptor` and its builders. I believe that `SerialKind`s don't have to have them

**Review by qwwdfsad — APPROVED** [review]
Great job!

One meta-comment: eventually we'll switch to `@sample` that will be auto-runnable from kotlinlang and from IDE. 
For that, you probably may want to change the sample form from
```
someFoo // computes to "kotlin.somefoo"
```

to the following
```
val someFooName = someFoo
print(someFoo) // Prints "kotlin.somefoo"
```

**Comment by sandwwraith** [comment]
Closed in favor of #2827

**Pull request closed without merging**